core/src/main/java/org/elasticsearch/index/query/InnerHitBuilder.java

```diff
@@ -565,17 +628,23 @@ public boolean equals(Object o) {
                 Objects.equals(sorts, that.sorts) &&
                 Objects.equals(highlightBuilder, that.highlightBuilder) &&
                 Objects.equals(query, that.query) &&
-                Objects.equals(innerHitsBuilder, that.innerHitsBuilder);
+                Objects.equals(childInnerHits, that.childInnerHits);
     }
 
     @Override
     public int hashCode() {
         return Objects.hash(name, nestedPath, parentChildType, from, size, explain, version, trackScores, fieldNames,
-                fieldDataFields, scriptFields, fetchSourceContext, sorts, highlightBuilder, query, innerHitsBuilder);
+                fieldDataFields, scriptFields, fetchSourceContext, sorts, highlightBuilder, query, childInnerHits);
     }
 
     public static InnerHitBuilder fromXContent(QueryParseContext context) throws IOException {
         return PARSER.parse(context.parser(), new InnerHitBuilder(), context);
     }
 
+    public static void extractInnerHits(QueryBuilder<?> query, Map<String, InnerHitBuilder> innerHitBuilders) {
+        if (query instanceof AbstractQueryBuilder) {
+            ((AbstractQueryBuilder) query).extractInnerHitBuilders(innerHitBuilders);
+        }
```

Comment:
can you throw an exception in the else clause, eg. "All queries must extend AbstractQueryBuilder but ..."